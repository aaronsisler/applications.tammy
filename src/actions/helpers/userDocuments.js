export const setUserDocuments = (userDocuments) => ({
    type: 'SET_USER_DOCUMENTS',
    userDocuments,
});

// export const addUserDocument = (userDocument) => ({
//     type: 'ADD_USER_DOCUMENT',
//     userDocument,
// });
