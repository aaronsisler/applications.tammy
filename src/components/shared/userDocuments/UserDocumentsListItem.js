import React from 'react';
import PropTypes from 'prop-types';

export default class UserDocumentsListItem extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <a className="document_list_item" href={`${this.props.downloadURL}`}>
                <div className="document_list_item__title">
                    {this.props.documentName}
                </div>
                <div className="document_list_item__date_uploaded">
                    {this.props.dateUploaded}
                </div>
            </a>
        );
    }
}

UserDocumentsListItem.propTypes = {
    downloadURL: PropTypes.string.isRequired,
    documentName: PropTypes.string.isRequired,
    dateUploaded: PropTypes.string.isRequired,
};
