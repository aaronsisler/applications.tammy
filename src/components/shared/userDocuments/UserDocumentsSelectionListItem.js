import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
    startAddApplicationUserDocument,
    startRemoveApplicationUserDocument,
} from 'Actions/application';

export class UserDocumentsSelectionListItem extends React.Component {
    constructor(props) {
        super(props);
    }

    handleUserDocumentSelection = () => {
        if (this.props.isDocumentAdded) {
            return this.props.startRemoveApplicationUserDocument(this.props.userDocumentId)
        }
        return this.props.startAddApplicationUserDocument(this.props.userDocumentId)
    }

    render() {
        const selectedClassName = this.props.isDocumentAdded ? "user_documents_selection_list_item_selected" : "";
        return (
            <div
                className={`user_documents_selection_list_item ${selectedClassName}`}
                onClick={this.handleUserDocumentSelection}
            >
                <div className="user_documents_selection_list_item__title">
                    {this.props.documentName}
                </div>
                <div className="user_documents_selection_list_item__date_uploaded">
                    {this.props.dateUploaded}
                </div>
            </div>
        );
    }
}

/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => ({
    startAddApplicationUserDocument: (userDocumentId) => dispatch(startAddApplicationUserDocument(userDocumentId)),
    startRemoveApplicationUserDocument: (userDocumentId) => dispatch(startRemoveApplicationUserDocument(userDocumentId)),
});

export default connect(undefined, mapDispatchToProps)(UserDocumentsSelectionListItem);


UserDocumentsSelectionListItem.propTypes = {
    dateUploaded: PropTypes.string.isRequired,
    documentName: PropTypes.string.isRequired,
    downloadUrl: PropTypes.string.isRequired,
    isDocumentAdded: PropTypes.bool.isRequired,
    startAddApplicationUserDocument: PropTypes.func.isRequired,
    startRemoveApplicationUserDocument: PropTypes.func.isRequired,
    userDocumentId: PropTypes.string.isRequired,
};
