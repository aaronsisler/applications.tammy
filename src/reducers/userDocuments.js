const userDocumentsReducerDefaultState = [];

const userDocumentsReducer = (state = userDocumentsReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_USER_DOCUMENTS':
            return action.userDocuments;
        default:
            return state;
    }
};

export default userDocumentsReducer;
