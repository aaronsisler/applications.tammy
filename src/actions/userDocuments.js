import database from 'Firebase/firebase';
import { addUserDocument, setUserDocuments } from 'Actions/helpers/userDocuments';

export const startSetUserDocuments = () => (dispatch, getState) => {
    const { uid: userId } = getState().auth;

    return database.ref(`user_documents/${userId}`).once('value').then((snapshot) => {
        const userDocuments = [];
        snapshot.forEach((childSnapshot) => {
            userDocuments.push({
                id: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        dispatch(setUserDocuments(userDocuments));
    });
};

export const startAddUserDocument = (userDocument) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    const { documentName, downloadUrl } = userDocument;
    const dateUploaded = new Date().toLocaleString();
    const uploadedDocument = { documentName, downloadUrl, dateUploaded }

    return database.ref(`user_documents/${userId}`)
        .push(uploadedDocument)
        .then((ref) => {
            dispatch(addUserDocument({
                id: ref.key,
                ...uploadedDocument
            }))
        });
}
