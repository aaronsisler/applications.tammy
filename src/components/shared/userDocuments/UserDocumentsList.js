import React from 'react';
import PropTypes from 'prop-types';
import UserDocumentsListItem from './UserDocumentsListItem';

export default class UserDocumentsList extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        if (this.props.userDocuments.length === 0) {
            return (
                <div className="user_documents_list empty">
                    No available items
                </div>
            );
        }

        return (
            <div className="user_documents_list">
                {
                    this.props.userDocuments.map((userDocument, index) =>
                        <UserDocumentsListItem key={index} {...userDocument} />)
                }
            </div>
        );
    }
}

UserDocumentsList.propTypes = {
    userDocuments: PropTypes.array,
};
