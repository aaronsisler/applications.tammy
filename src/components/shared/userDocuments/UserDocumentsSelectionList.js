import React from 'react';
import PropTypes from 'prop-types';
import UserDocumentsSelectionListItem from './UserDocumentsSelectionListItem';

export default class UserDocumentsSelectionList extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        if (this.props.userDocuments.length === 0) {
            return (
                <div className="user_documents_selection_list empty">
                    No available items
                </div>
            );
        }

        const applicationUserDocumentIds = (
            this.props.applicationUserDocuments.map((applicationUserDocument) => applicationUserDocument.userDocumentId)
        );

        return (
            <div className="user_documents_selection_list">
                {
                    this.props.userDocuments.map((userDocument, index) => {
                        const isDocumentAdded = applicationUserDocumentIds
                            .indexOf(userDocument.userDocumentId) != -1;

                        return (
                            <UserDocumentsSelectionListItem
                                key={index}
                                isDocumentAdded={isDocumentAdded}
                                {...userDocument}
                            />
                        );
                    })
                }
            </div>
        );
    }
}

UserDocumentsSelectionList.propTypes = {
    applicationUserDocuments: PropTypes.array.isRequired,
    userDocuments: PropTypes.array.isRequired,
};
