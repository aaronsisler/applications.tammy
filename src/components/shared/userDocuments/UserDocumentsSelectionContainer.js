import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import UserDocumentsSelectionList from './UserDocumentsSelectionList';

export class UserDocumentsSelectionContainer extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div className="user_documents_selection_container">
                <UserDocumentsSelectionList
                    applicationUserDocuments={this.props.applicationUserDocuments}
                    userDocuments={this.props.userDocuments}
                />
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = (state) => ({
    applicationUserDocuments: state.application.userDocuments,
    userDocuments: state.userDocuments,
});

export default connect(mapStateToProps)(UserDocumentsSelectionContainer);

UserDocumentsSelectionContainer.propTypes = {
    applicationUserDocuments: PropTypes.array.isRequired,
    userDocuments: PropTypes.array.isRequired,
};
