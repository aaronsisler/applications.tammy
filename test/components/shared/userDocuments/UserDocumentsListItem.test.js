import React from 'react';
import { shallow } from 'enzyme';
import UserDocumentsListItem from 'Shared/userDocuments/UserDocumentsListItem';
import userDocuments from '../../../fixtures/userDocuments';

describe('UserDocumentsListItem', () => {
    const [userDocument] = userDocuments;
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <UserDocumentsListItem
                {...userDocument}
            />);
    });

    it('should render UserDocumentsListItem correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
