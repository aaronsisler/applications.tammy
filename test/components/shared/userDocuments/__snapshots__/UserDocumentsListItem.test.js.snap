// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDocumentsListItem should render UserDocumentsListItem correctly 1`] = `
<a
  className="document_list_item"
  href="downloadUrl1"
>
  <div
    className="document_list_item__title"
  >
    Document Name 1
  </div>
  <div
    className="document_list_item__date_uploaded"
  >
    2018/01/01 01:23:45A
  </div>
</a>
`;
